File: 921-MinimumAddtoMakeParenthesesValid/921-MinimumAddtoMakeParenthesesValid.java
/*

921. Minimum Add to Make Parentheses Valid
Medium

Given a string S of '(' and ')' parentheses, we add the minimum number of parentheses ( '(' or ')', and in any positions ) so that the resulting parentheses string is valid.

Formally, a parentheses string is valid if and only if:

    It is the empty string, or
    It can be written as AB (A concatenated with B), where A and B are valid strings, or
    It can be written as (A), where A is a valid string.

Given a parentheses string, return the minimum number of parentheses we must add to make the resulting string valid.

Example 1:

Input: "())"
Output: 1

Example 2:

Input: "((("
Output: 3

Example 3:

Input: "()"
Output: 0

Example 4:

Input: "()))(("
Output: 4

Note:

    S.length <= 1000
    S only consists of '(' and ')' characters.

*/

class Solution {
    public int minAddToMakeValid(String S) {
        Map<Character, Character> map = new HashMap<>();
        map.put('(', ')');
        Stack<Character> stack = new Stack<>();
        int count = 0;
        for(char ch : S.toCharArray()){
            if(map.containsKey(ch))
                stack.push(map.get(ch));
            else{
                if(!stack.isEmpty())
                    stack.pop();
                else
                    count++;
            }
        }
        return stack.size() + count;
    }
}

File: 921-MinimumAddtoMakeParenthesesValid/921-MinimumAddtoMakeParenthesesValid.py
"""

921. Minimum Add to Make Parentheses Valid

A parentheses string is valid if and only if:

    It is the empty string,
    It can be written as AB (A concatenated with B), where A and B are valid strings, or
    It can be written as (A), where A is a valid string.

You are given a parentheses string s. In one move, you can insert a parenthesis at any position of the string.

    For example, if s = "()))", you can insert an opening parenthesis to be "(()))" or a closing parenthesis to be "())))".

Return the minimum number of moves required to make s valid.

Example 1:

Input: s = "())"
Output: 1

Example 2:

Input: s = "((("
Output: 3

Constraints:

    1 <= s.length <= 1000
    s[i] is either '(' or ')'.

"""

class Solution:
    def minAddToMakeValid(self, s: str) -> int:
        mapp = {'(':')'}
        stack = []
        res = 0
        for ch in s:
            if ch in mapp:
                stack.append(mapp[ch])
            else:
                if not stack: res += 1
                else: stack.pop()
        return res + len(stack)
